src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -118,6 +271,166 @@ int16_t LoRaWANNode::setBufferNonces(uint8_t* persistentBuffer) {
+  if(this->band->bandType == RADIOLIB_LORAWAN_BAND_DYNAMIC) {
+    this->selectChannelPlanDyn(true);
+  } else {
+    this->selectChannelPlanFix();
+  }
+
+  uint8_t drUp = RADIOLIB_LORAWAN_DATA_RATE_UNUSED;
+
+  // on fixed bands, the first OTAA uplink (JoinRequest) is sent on fixed datarate
+  if(this->band->bandType == RADIOLIB_LORAWAN_BAND_FIXED && lwMode == RADIOLIB_LORAWAN_MODE_OTAA) {
+    // randomly select one of 8 or 9 channels and find corresponding datarate
+    uint8_t numChannels = this->band->numTxSpans == 1 ? 8 : 9;
+    uint8_t rand = this->phyLayer->random(numChannels) + 1;     // range 1-8 or 1-9
+    if(rand <= 8) {
+      drUp = this->band->txSpans[0].drJoinRequest;        // if one of the first 8 channels, select datarate of span 0
+    } else {
+      drUp = this->band->txSpans[1].drJoinRequest;        // if ninth channel, select datarate of span 1
+    }
+  }
+
+  // on dynamic bands, the first OTAA uplink (JoinRequest) can be any available datarate
+  // this is also true for ABP where there is no JoinRequest
+  if(initialDr != RADIOLIB_LORAWAN_DATA_RATE_UNUSED) {
+    uint8_t i = 0; 
+    for(; i < RADIOLIB_LORAWAN_NUM_AVAILABLE_CHANNELS; i++) {
+      if(this->channelPlan[RADIOLIB_LORAWAN_UPLINK][i].enabled) {
+        if(initialDr >= this->channelPlan[RADIOLIB_LORAWAN_UPLINK][i].drMin
+          && initialDr <= this->channelPlan[RADIOLIB_LORAWAN_UPLINK][i].drMax) {
+            break;
+        }
+      }
+    }
+    // if there is no channel that allowed the user-specified datarate, revert to default datarate
+    if(i == RADIOLIB_LORAWAN_NUM_AVAILABLE_CHANNELS) {
+      RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Datarate %d is not valid - using default", initialDr);
+      initialDr = RADIOLIB_LORAWAN_DATA_RATE_UNUSED;
+    }
+  }
+
+  // if there is no (channel that allowed the) user-specified datarate, use a default datarate
+  // we use the floor of the average datarate of the first enabled channel
+  if(initialDr == RADIOLIB_LORAWAN_DATA_RATE_UNUSED) {
+    for(int i = 0; i < RADIOLIB_LORAWAN_NUM_AVAILABLE_CHANNELS; i++) {
+      if(this->channelPlan[RADIOLIB_LORAWAN_UPLINK][i].enabled) {
+        uint8_t drMin = this->channelPlan[RADIOLIB_LORAWAN_UPLINK][i].drMin;
+        uint8_t drMax = this->channelPlan[RADIOLIB_LORAWAN_UPLINK][i].drMax;
+        drUp = (drMin + drMax) / 2;
+      }
+    }
+  }
+
+  uint8_t cOcts[5];                       // 5 = maximum downlink payload length
+  uint8_t cid = RADIOLIB_LORAWAN_MAC_LINK_ADR;
+  uint8_t cLen = 1;                       // only apply Dr/Tx field
+  cOcts[0]  = (drUp << 4);                // set uplink datarate
+  cOcts[0] |= 0;                          // default to max Tx Power
+  (void)execMacCommand(cid, cOcts, cLen);
+
+  cid = RADIOLIB_LORAWAN_MAC_DUTY_CYCLE;
+  this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  uint8_t maxDCyclePower;
+  switch(this->band->dutyCycle) {
+    case(0):
+      maxDCyclePower = 0;
+      break;
```

Comment:
If we do not want to error on an "invalid" value of `this->band->dutyCycle`, then this can be simplified by initializing `maxDCyclePower = 0` and then removing both `case(0)` as well as `default`.